include/swift/AST/TypeRepr.h

```diff
@@ -572,109 +561,123 @@ class ImplicitlyUnwrappedOptionalTypeRepr : public TypeRepr {
   friend class TypeRepr;
 };
 
+/// \brief A parsed element within a tuple type.
+struct TupleTypeReprElement {
+  Identifier Name;
+  SourceLoc NameLoc;
+  Identifier SecondName;
+  SourceLoc SecondNameLoc;
+  SourceLoc UnderscoreLoc;
+  SourceLoc ColonLoc;
+  SourceLoc InOutLoc;
+  TypeRepr *Type;
+  SourceLoc TrailingCommaLoc;
+
+  TupleTypeReprElement() {}
+  TupleTypeReprElement(TypeRepr *Type): Type(Type) {}
+};
+
 /// \brief A tuple type.
 /// \code
 ///   (Foo, Bar)
 ///   (x: Foo)
 ///   (_ x: Foo)
 /// \endcode
 class TupleTypeRepr final : public TypeRepr,
-    private llvm::TrailingObjects<TupleTypeRepr, TypeRepr*, Identifier,
-                                  SourceLoc, std::pair<SourceLoc, unsigned>> {
+    private llvm::TrailingObjects<TupleTypeRepr, TupleTypeReprElement> {
   friend TrailingObjects;
-  typedef std::pair<SourceLoc, unsigned> SourceLocAndIdx;
 
-  unsigned NumElements;
-  SourceRange Parens;
-
-  enum {
+  enum NameInfo: uint8_t {
```

Comment:
Oh, it's the AST *dumper* that checks `hasElementNames()`. Let's just kill the bit---it's fine to have an O(n) check for "has any element names". I do think it'd be great for `NumElements` to go into `TupleTypeReprBitfields`.